pipeline_dp/budget_accounting.py

```diff
@@ -100,7 +101,188 @@ def compute_budgets(self):
+            standard deviation until compute_budgets is called.
+        """
+        mechanism_spec = MechanismSpec(noise_kind=noise_kind)
+        mechanism_spec_internal = MechanismSpecInternal(
+            mechanism_spec=mechanism_spec)
+        self._mechanisms.append(mechanism_spec_internal)
+        return mechanism_spec
+
+    def compute_budgets(self):
+        """Computes the budget for the pipeline.
+
+        Composes the mechanisms and adjusts the amount of
+        noise based on given epsilon. Sets the noise for the
+        entire pipeline.
+        """
+        if not self._mechanisms:
+            return
+        minimum_noise_std = self._find_minimum_noise_std()
+        self.minimum_noise_std = minimum_noise_std
+        for mechanism in self._mechanisms:
+            mechanism.mechanism_spec._noise_standard_deviation = minimum_noise_std / mechanism.weight
+
+    def _find_minimum_noise_std(self) -> float:
+        """Find a noise value to the minimum.
+
+        Use binary search to find a minimum noise value that gives a
+        new epsilon close to the given epsilon (within a threshold).
+        By increasing the noise we can decrease the epsilon.
+
+        Returns:
+            The noise value adjusted for the given epsilon.
+        """
+        threshold = 1e-4
+        maximum_noise_std = self._calculate_max_noise_std()
+        low, high = 0, maximum_noise_std
+        while low + threshold < high:
+            mid = (high - low) / 2 + low
+            pld = self._compose_distributions(mid)
+            pld_epsilon = pld.get_epsilon_for_delta(self._total_delta)
+            if pld_epsilon <= self._total_epsilon:
+                high = mid
+            elif pld_epsilon > self._total_epsilon:
+                low = mid
+
+        return high
+
+    def _calculate_max_noise_std(self) -> float:
+        """Calculate maximum noise std for given epsilon.
+
+        Returns:
+            The maximum noise std value adjusted for the given epsilon.
+        """
+        max_noise_std = 1
+        pld_epsilon = self._total_epsilon + 1
+        while pld_epsilon > self._total_epsilon:
+            max_noise_std *= 2
+            pld = self._compose_distributions(max_noise_std)
+            pld_epsilon = pld.get_epsilon_for_delta(self._total_delta)
+        return max_noise_std
+
+    def _compose_distributions(
+            self, noise_std: float) -> pldlib.PrivacyLossDistribution:
+        """ Uses the Privacy Loss Distribution library to compose distributions.
+
+        Args:
+            noise_std: The noise of the distributions to construct.
+
+        Returns:
+            A PrivacyLossDistribution object for the pipeline.
+        """
+        composed, pld = None, None
+
+        for mechanism_spec_internal in self._mechanisms:
+            if mechanism_spec_internal.mechanism_spec.noise_kind == NoiseKind.LAPLACE:
+                pld = pldlib.PrivacyLossDistribution.from_laplace_mechanism(
```

Comment:
Please add a comment before:
\# The Laplace distribution parameter = std/sqrt(2).